pkg/apis/migration/migration.go

```diff
@@ -0,0 +1,351 @@
+
+// VolumeMigrationService exposes interfaces to manage VolumeInfo cache and CR
+// for VolumePath, VolumeID and VolumeName mapping
+type VolumeMigrationService interface {
+	// GetVolumeInfo returns CnsvSphereVolumeMigrationSpec for given VolumePath from local cache
+	// if CnsvSphereVolumeMigrationSpec is not available in the cache, K8s CR look up will be performed, if that too
+	// can not retrieve VolumeInfo, ErrVolumeNotRegistered will be returned
+	GetVolumeInfo(ctx context.Context, volumePath string) (migration.CnsvSphereVolumeMigrationSpec, error)
+
+	// SaveVolumeInfo helps create CR for given cnsvSphereVolumeMigration
+	// this func also update local cache with supplied cnsvSphereVolumeMigration, after successful creation of CR
+	SaveVolumeInfo(ctx context.Context, cnsvSphereVolumeMigration migration.CnsvSphereVolumeMigration) error
+
+	// DeleteVolumeInfo helps delete cached mapping of volumePath to VolumeInfo and
+	// also helps delete K8s Custom Resource for specified volumePath
+	DeleteVolumeInfo(ctx context.Context, volumePath string) error
+
+	// LoadAllVolumeInfo helps cache all CR for CnsvSphereVolumeMigration
+	LoadAllVolumeInfo(ctx context.Context) error
+
+	// Register Volume takes in-tree VolumePath and helps register Volume with CNS
+	RegisterVolume(ctx context.Context, volumePath string, manager *common.Manager) (migration.CnsvSphereVolumeMigration, error)
+}
+
+// VolumeMigration holds migrated volume information and provides functionality around it.
+type volumeMigration struct {
+	// Volume Path to volumeInfo map
+	volumePathToVolumeInfo sync.Map
+	// migrationClient helps operate on CnsvSphereVolumeMigration custom resource
+	migrationClient client.Client
+}
+
+// onceForInTreeVolumeMigrationService is used for initializing the VolumeMigrationService singleton.
+var onceForVolumeMigrationService sync.Once
+
+// volumeMigrationInstance is instance of volumeMigration and implements interface for VolumeMigrationService
+var volumeMigrationInstance *volumeMigration
+
+// GetVolumeMigrationService returns the singleton VolumeMigrationService
+func GetVolumeMigrationService(ctx context.Context) (VolumeMigrationService, error) {
+	var err error
+	log := logger.GetLogger(ctx)
+	onceForVolumeMigrationService.Do(func() {
+		log.Info("Initializing volume migration service...")
+		volumeMigrationInstance = &volumeMigration{
+			volumePathToVolumeInfo: sync.Map{},
+		}
+		err = createrCnsvSphereVolumeMigrationCRD(ctx)
+		if err != nil {
+			log.Errorf("failed to create volume migration CRD. Error: %v", err)
+			return
+		}
+		config, err := k8s.GetKubeConfig(ctx)
+		if err != nil {
+			log.Errorf("failed to get kubeconfig. err: %v", err)
+			return
+		}
+		volumeMigrationInstance.migrationClient, err = k8s.NewClientForGroup(ctx, config, "cns.vmware.com")
+		if err != nil {
+			log.Errorf("failed to create migrationClient. Err: %v", err)
+			return
+		}
+		log.Info("volume migration service initialized")
```

Comment:
You could load all CRs as part of creating volumeMigrationInstance and avoid exposing `LoadAllVolumeInfo` as a public function.

Reply:
yes doing that.